cmd/snap-bootstrap/inputwatch/evdev.go

```diff
@@ -0,0 +1,139 @@
+// -*- Mode: Go; indent-tabs-mode: t -*-
+
+/*
+ * Copyright (C) 2020 Canonical Ltd
+ *
+ * This program is free software: you can redistribute it and/or modify
+ * it under the terms of the GNU General Public License version 3 as
+ * published by the Free Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ *
+ */
+
+package inputwatch
+
+import (
+	"fmt"
+
+	// TODO:UC20: not packaged, reimplment the minimal things we need?
+	evdev "github.com/gvalkov/golang-evdev"
+
+	"github.com/snapcore/snapd/logger"
+)
+
+var (
+	evKeyCapability = evdev.CapabilityType{Type: evdev.EV_KEY, Name: "EV_KEY"}
+
+	// chooserTriggerKey is what activates the chooser when held down
+	chooserTriggerKey = evdev.CapabilityCode{Code: 2, Name: "KEY_1"}
+)
+
+func init() {
+	input = &evdevInput{}
+}
+
+type evdevKeyboardInputDevice struct {
+	keyCode uint16
+	dev     *evdev.InputDevice
+}
+
+func (e *evdevKeyboardInputDevice) WaitForTrigger(ch chan keyEvent) {
+	const triggerHeldCount = 20
+
+	heldCount := uint(0)
+
+	// TODO: find a reliable way to get/set key repeat rate? EVIOCGREP seems
+	// to return bad info; the support varies from driver to driver,
+	// gpio-keys may require 'autorepeat' enabled at the devicetree level
+	e.dev.SetRepeatRate(1000, 0)
```

Comment:
Is this restored later to whatever is the default for this device?

Reply:
I suppose the programs that use input should set the repeat rate themselves, but we could probably do it in a defer.

FWIW I tried adding the calls to GetRepeatRate() and SetRepeatRate() with original values in a defer but I could not say whether those were effective as there's no error return from SetRepeatRate() :disappointed:

Reply:
from reading some old articles it seems like the device capabilities should tell if repeat can be set/get on it

Reply:
From what I've seen, the capabilities reported back by go-evdev do not contain an EV_REP type. I haven't seen a way to tell whether a device even generates a KeyHold event. For instance, in a VM where an AT keyboard is emulated I see key up/down/hold events. Then on my USB keyboard, I only see down and up.